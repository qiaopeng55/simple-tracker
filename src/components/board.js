import React, { Component } from 'react';

import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import { connect } from 'react-redux';
import './main.css';
import { BoardContainer, ColumnContainer } from './styles';
import Column from './column';

const Columns = ({ columns, cards }) =>
  columns.map(column => (
    <Column
      key={column.id}
      {...column}
      cards={cards.filter(card => card.columnId === column.id)}
    />
  ));

export class Board extends Component {
  render() {
    const { columns, cards } = this.props;
    return (
      <BoardContainer>
        <div id="header">Simple Issue Tracker</div>
        <ColumnContainer>
          {Columns({
            columns,
            cards,
          })}
        </ColumnContainer>
      </BoardContainer>
    );
  }
}

const mapStateToProps = state => ({
  columns: state.columns,
  cards: state.cards,
});

export default connect(mapStateToProps)(DragDropContext(HTML5Backend)(Board));
