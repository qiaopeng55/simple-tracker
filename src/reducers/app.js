/**
 *  App reducer
 *  manage key app state here
 */

// Import Action Types
import { ADDING_CARD, EDITING_CARD } from '../actions/constant';

function App(state = {}, action) {
  switch (action.type) {
    case ADDING_CARD:
      return {
        operation: ADDING_CARD,
      };
    case EDITING_CARD:
      return {
        operation: EDITING_CARD,
      };
    default:
      return state;
  }
}

export default App;
