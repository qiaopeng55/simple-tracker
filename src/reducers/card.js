/**
 *  card reducer
 */

import uuidv1 from 'uuid/v1';
// Import Action Types
import {
  ADDING_CARD,
  UPDATE_CARD,
  SAVING_CARD,
  CLEAR_CARD,
  EDITING_CARD,
  EDITING_CARD_SAVE,
} from '../actions/constant';

function card(state = {}, action) {
  switch (action.type) {
    case ADDING_CARD:
      return {
        id: uuidv1(),
        columnId: action.data.columnId,
      };
    case UPDATE_CARD:
      return {
        ...action.data.card,
      };
    case SAVING_CARD:
    case CLEAR_CARD:
      return {};
    case EDITING_CARD:
      return {
        ...action.data.card,
      };
    case EDITING_CARD_SAVE:
      return {};
    default:
      return state;
  }
}

export default card;
