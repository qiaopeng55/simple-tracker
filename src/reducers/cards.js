import swapArrayItems from '../utils/swap-array-items';
import getCardIndex from '../utils/get-card-index';

import {
  DROP_CARD_TO_COLUMN,
  DROP_CARD_ON_CARD,
  SAVING_CARD,
  DELETE_CARD,
  EDITING_CARD_SAVE,
} from '../actions/constant';

/**
 *  columns reducer
 */

export const cardsInitialState = [
  {
    id: 1,
    text: 'task 1',
    columnId: 1,
  },
  {
    id: 2,
    text: 'task 2',
    columnId: 1,
  },
  {
    id: 3,
    text: 'task 3',
    columnId: 1,
  },
  {
    id: 4,
    text: 'task 4',
    columnId: 2,
  },
  {
    id: 5,
    text: 'task 5',
    columnId: 2,
  },
  {
    id: 6,
    text: 'task 6',
    columnId: 2,
  },
  {
    id: 7,
    text: 'task 7',
    columnId: 2,
  },
  {
    id: 8,
    text: 'task 8',
    columnId: 3,
  },
  {
    id: 9,
    text: 'task 9',
    columnId: 3,
  },
];

function columns(state = cardsInitialState, action) {
  switch (action.type) {
    case DROP_CARD_TO_COLUMN:
      return state.map(
        card =>
          card.id === action.data.cardId
            ? {
                ...card,
                columnId: action.data.columnId,
              }
            : card
      );
    case DROP_CARD_ON_CARD:
      return swapArrayItems(
        state,
        getCardIndex(state, action.data.cardId),
        getCardIndex(state, action.data.targetCardId)
      );
    case SAVING_CARD:
      return [...state, action.data.card];
    case DELETE_CARD:
      return state.filter(card => card.id !== action.data.cardId);
    case EDITING_CARD_SAVE:
      return state.map(
        card =>
          card.id === action.data.card.id
            ? {
                ...action.data.card,
              }
            : card
      );
    default:
      return state;
  }
}

export default columns;
