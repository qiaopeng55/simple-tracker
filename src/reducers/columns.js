/**
 *  columns reducer
 */

export const columnsInitialState = [
  {
    id: 1,
    title: 'todo',
  },
  {
    id: 2,
    title: 'in progress',
  },
  {
    id: 3,
    title: 'done',
  },
];

function columns(state = columnsInitialState, action) {
  switch (action.type) {
    default:
      return state;
  }
}

export default columns;
