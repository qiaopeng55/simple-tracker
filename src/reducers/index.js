import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';
import columns from './columns';
import cards from './cards';
import card from './card';
import app from './app';

export default combineReducers({
  app,
  columns,
  cards,
  card,
  router: routerReducer,
});
