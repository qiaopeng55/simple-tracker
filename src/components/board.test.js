import React from 'react';
import { shallow } from 'enzyme';
import { columnsInitialState } from '../reducers/columns';
import { cardsInitialState } from '../reducers/cards';
import { Board } from './board';

const props = {
  columns: columnsInitialState,
  cards: cardsInitialState,
};

describe('Board', () => {
  const board = shallow(<Board {...props} />);

  it('renders properly', () => {
    expect(board).toMatchSnapshot();
  });

  it('renders the title', () => {
    expect(board.find('#header').text()).toEqual('Simple Issue Tracker');
  });

  it('renders correct number of columns', () => {
    expect(board.find('Connect(DropTarget(Column))').length).toEqual(
      columnsInitialState.length
    );
  });

  it('pass correct title to Column component', () => {
    board.find('Connect(DropTarget(Column))').forEach((col, index) => {
      expect(col.props().title).toEqual(columnsInitialState[index].title);
    });
  });

  it('pass cards with matching column id to Column component', () => {
    board.find('Connect(DropTarget(Column))').forEach(col => {
      col.props().cards.forEach(card => {
        expect(card.columnId).toEqual(col.props().id);
      });
    });
  });
});
