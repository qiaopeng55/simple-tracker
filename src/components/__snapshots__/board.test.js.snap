// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders properly 1`] = `
<styled.div>
  <div
    id="header"
  >
    Simple Issue Tracker
  </div>
  <styled.div>
    <Connect(DropTarget(Column))
      cards={
        Array [
          Object {
            "columnId": 1,
            "id": 1,
            "text": "task 1",
          },
          Object {
            "columnId": 1,
            "id": 2,
            "text": "task 2",
          },
          Object {
            "columnId": 1,
            "id": 3,
            "text": "task 3",
          },
        ]
      }
      id={1}
      key="1"
      title="todo"
    />
    <Connect(DropTarget(Column))
      cards={
        Array [
          Object {
            "columnId": 2,
            "id": 4,
            "text": "task 4",
          },
          Object {
            "columnId": 2,
            "id": 5,
            "text": "task 5",
          },
          Object {
            "columnId": 2,
            "id": 6,
            "text": "task 6",
          },
          Object {
            "columnId": 2,
            "id": 7,
            "text": "task 7",
          },
        ]
      }
      id={2}
      key="2"
      title="in progress"
    />
    <Connect(DropTarget(Column))
      cards={
        Array [
          Object {
            "columnId": 3,
            "id": 8,
            "text": "task 8",
          },
          Object {
            "columnId": 3,
            "id": 9,
            "text": "task 9",
          },
        ]
      }
      id={3}
      key="3"
      title="done"
    />
  </styled.div>
</styled.div>
`;
